[BUG] kyverno initContainer fail to start when external.metrics.k8s.io/v1beta1 return empty array for resources

**Software version numbers**
State the version numbers of applications involved in the bug.

* Kubernetes version: 1.16
* Kyverno version: v1.3.0 and v1.3.1

**Describe the bug**
A clear and concise description of what the bug is.
Kyverno init container fail to start with external metrics enable

**To Reproduce**
Steps to reproduce the behavior:
1. install prometheus external metrics
2. install kyverno v1.3.0 or v1.3.1
3. kyverno pod status is Init:CrashLoopBackOff
4. See error
``I0121 23:21:03.940127       1 main.go:140]  "msg"="Using in-cluster configuration"
E0121 23:21:03.967938       1 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
E0121 23:21:03.968323       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.968343       1 client.go:286] dclient "msg"="schema not found"  "kind"="ValidatingWebhookConfiguration"
E0121 23:21:03.968727       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.968755       1 client.go:286] dclient "msg"="schema not found"  "kind"="ValidatingWebhookConfiguration"
E0121 23:21:03.970418       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.970443       1 client.go:286] dclient "msg"="schema not found"  "kind"="MutatingWebhookConfiguration"
E0121 23:21:03.970578       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.970592       1 client.go:286] dclient "msg"="schema not found"  "kind"="MutatingWebhookConfiguration"
E0121 23:21:03.971480       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.971498       1 client.go:286] dclient "msg"="schema not found"  "kind"="ValidatingWebhookConfiguration"
E0121 23:21:03.971788       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.971799       1 client.go:286] dclient "msg"="schema not found"  "kind"="ValidatingWebhookConfiguration"
E0121 23:21:03.972535       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.972554       1 client.go:286] dclient "msg"="schema not found"  "kind"="MutatingWebhookConfiguration"
E0121 23:21:03.973374       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.973395       1 client.go:286] dclient "msg"="schema not found"  "kind"="MutatingWebhookConfiguration"
E0121 23:21:03.973791       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.973804       1 client.go:286] dclient "msg"="schema not found"  "kind"="Namespace"
E0121 23:21:03.974495       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.974519       1 client.go:286] dclient "msg"="schema not found"  "kind"="ClusterPolicyReport"
E0121 23:21:03.974961       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.974996       1 client.go:286] dclient "msg"="schema not found"  "kind"="ReportChangeRequest"
E0121 23:21:03.975570       1 client.go:337] dclient "msg"="failed to get registered preferred resources" "error"="unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1"
I0121 23:21:03.975586       1 client.go:286] dclient "msg"="schema not found"  "kind"="ClusterReportChangeRequest"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x115b350]

goroutine 25 [running]:
main.removeReportChangeRequest(0xc00007f220, 0x13bfe86, 0x13, 0xc000082901, 0xc000397ef0)
	/home/runner/work/kyverno/kyverno/cmd/initContainer/main.go:312 +0x110
main.executeRequest(0xc00007f220, 0x13bfe86, 0x13, 0x13b48e8, 0x0, 0x0, 0x0)
	/home/runner/work/kyverno/kyverno/cmd/initContainer/main.go:128 +0x179
main.process.func1(0xc0000829c0, 0xc000082960, 0xc00007f220, 0xc000082900, 0x158b720, 0xc000325ec0, 0xc0000825a0)
	/home/runner/work/kyverno/kyverno/cmd/initContainer/main.go:188 +0xfb
created by main.process
	/home/runner/work/kyverno/kyverno/cmd/initContainer/main.go:184 +0xd7`
`
5. `kubectl` get --raw "/apis/external.metrics.k8s.io/v1beta1" | jq
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "external.metrics.k8s.io/v1beta1",
  "resources": []
}`
6. delete external metrics api, redeploy kyverno, it starts successful.



**Expected behavior**
A clear and concise description of what you expected to happen.
external metrics return empty resource which is correct when there is no metrics. Kyverno shouldn't fail with empty resource.

--- a/cmd/initContainer/main.go
+++ b/cmd/initContainer/main.go
@@ -46,6 +46,7 @@ func main() {
 	if err := flag.Set("v", "2"); err != nil {
 		klog.Fatalf("failed to set log level: %v", err)
 	}
+
 	flag.Parse()
 
 	// os signal handler
@@ -109,9 +110,10 @@ func main() {
 			log.Log.Error(err, "failed to cleanup resource")
 		}
 	}
+
 	// if there is any failure then we fail process
 	if failure {
-		log.Log.Info("failed to cleanup webhook configurations")
+		log.Log.Info("failed to cleanup prior configurations")
 		os.Exit(1)
 	}
 }
@@ -131,6 +133,7 @@ func executeRequest(client *client.Client, req request) error {
 	case policyViolation, clusterPolicyViolation:
 		return removeViolationCRD(client)
 	}
+
 	return nil
 }
 
@@ -140,6 +143,7 @@ func createClientConfig(kubeconfig string) (*rest.Config, error) {
 		logger.Info("Using in-cluster configuration")
 		return rest.InClusterConfig()
 	}
+
 	logger.Info(fmt.Sprintf("Using configuration from '%s'", kubeconfig))
 	return clientcmd.BuildConfigFromFlags("", kubeconfig)
 }

--- a/cmd/kyverno/main.go
+++ b/cmd/kyverno/main.go
@@ -123,7 +123,7 @@ func main() {
 
 	// CRD CHECK
 	// - verify if Kyverno CRDs are available
-	if !utils.CRDInstalled(client.DiscoveryClient, log.Log) {
+	if !utils.CRDsInstalled(client.DiscoveryClient) {
 		setupLog.Error(fmt.Errorf("CRDs not installed"), "Failed to access Kyverno CRDs")
 		os.Exit(1)
 	}

--- a/pkg/dclient/client.go
+++ b/pkg/dclient/client.go
@@ -147,6 +147,7 @@ func (c *Client) ListResource(apiVersion string, kind string, namespace string, 
 	if lselector != nil {
 		options = meta.ListOptions{LabelSelector: helperv1.FormatLabelSelector(lselector)}
 	}
+
 	return c.getResourceInterface(apiVersion, kind, namespace).List(context.TODO(), options)
 }
 
@@ -325,31 +326,38 @@ func (c ServerPreferredResources) FindResource(apiVersion string, kind string) (
 }
 
 func (c ServerPreferredResources) findResource(apiVersion string, kind string) (*meta.APIResource, schema.GroupVersionResource, error) {
-	var serverresources []*meta.APIResourceList
+	var serverResources []*meta.APIResourceList
 	var err error
 	if apiVersion == "" {
-		serverresources, err = c.cachedClient.ServerPreferredResources()
+		serverResources, err = c.cachedClient.ServerPreferredResources()
 	} else {
-		serverresources, err = c.cachedClient.ServerResources()
+		_, serverResources, err = c.cachedClient.ServerGroupsAndResources()
 	}
 
 	if err != nil {
-		c.log.Error(err, "failed to get registered preferred resources")
-		return nil, schema.GroupVersionResource{}, err
+		if discovery.IsGroupDiscoveryFailedError(err) {
+			logDiscoveryErrors(err, c)
+		} else if isMetricsServerUnavailable(kind, err) {
+			c.log.V(3).Info("failed to find preferred resource version", "error", err.Error())
+		} else {
+			c.log.Error(err, "failed to find preferred resource version")
+			return nil, schema.GroupVersionResource{}, err
+		}
 	}
 
-	for _, serverresource := range serverresources {
+	for _, serverResource := range serverResources {
 
-		if apiVersion != "" && serverresource.GroupVersion != apiVersion {
+		if apiVersion != "" && serverResource.GroupVersion != apiVersion {
 			continue
 		}
-		for _, resource := range serverresource.APIResources {
+
+		for _, resource := range serverResource.APIResources {
 
 			// skip the resource names with "/", to avoid comparison with subresources
 			if resource.Kind == kind && !strings.Contains(resource.Name, "/") {
-				gv, err := schema.ParseGroupVersion(serverresource.GroupVersion)
+				gv, err := schema.ParseGroupVersion(serverResource.GroupVersion)
 				if err != nil {
-					c.log.Error(err, "failed to parse groupVersion", "groupVersion", serverresource.GroupVersion)
+					c.log.Error(err, "failed to parse groupVersion", "groupVersion", serverResource.GroupVersion)
 					return nil, schema.GroupVersionResource{}, err
 				}
 
@@ -359,4 +367,25 @@ func (c ServerPreferredResources) findResource(apiVersion string, kind string) (
 	}
 
 	return nil, schema.GroupVersionResource{}, fmt.Errorf("kind '%s' not found in apiVersion '%s'", kind, apiVersion)
+}
+
+func logDiscoveryErrors(err error, c ServerPreferredResources) {
+	discoveryError := err.(*discovery.ErrGroupDiscoveryFailed)
+	for gv, e := range discoveryError.Groups {
+		if gv.Group == "custom.metrics.k8s.io" || gv.Group == "metrics.k8s.io" {
+			// These error occur when Prometheus is installed as an external metrics server
+			// See: https://github.com/kyverno/kyverno/issues/1490
+			c.log.V(3).Info("failed to retrieve metrics API group", "gv", gv)
+			continue
+		}
+
+		c.log.Error(e, "failed to retrieve API group", "gv", gv)
+	}
+}
+
+func isMetricsServerUnavailable(kind string, err error) bool {
+	// error message is defined at:
+	// https://github.com/kubernetes/apimachinery/blob/2456ebdaba229616fab2161a615148884b46644b/pkg/api/errors/errors.go#L432
+	return strings.HasPrefix(kind, "metrics.k8s.io/") &&
+		strings.Contains(err.Error(), "the server is currently unable to handle the request")
 }

--- a/pkg/generate/generate_controller.go
+++ b/pkg/generate/generate_controller.go
@@ -3,8 +3,6 @@ package generate
 import (
 	"time"
 
-	"sigs.k8s.io/controller-runtime/pkg/log"
-
 	"github.com/go-logr/logr"
 	kyverno "github.com/kyverno/kyverno/pkg/api/kyverno/v1"
 	kyvernoclient "github.com/kyverno/kyverno/pkg/client/clientset/versioned"
@@ -276,8 +274,6 @@ func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
 // worker runs a worker thread that just dequeues items, processes them, and marks them done.
 // It enforces that the syncHandler is never invoked concurrently with the same key.
 func (c *Controller) worker() {
-	log.Log.Info("starting new worker...")
-
 	for c.processNextWorkItem() {
 	}
 }

--- a/pkg/utils/util.go
+++ b/pkg/utils/util.go
@@ -2,10 +2,9 @@ package utils
 
 import (
 	"fmt"
-	"reflect"
 	"regexp"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 	"strconv"
-	"strings"
 
 	"github.com/go-logr/logr"
 	client "github.com/kyverno/kyverno/pkg/dclient"
@@ -57,35 +56,30 @@ func NewKubeClient(config *rest.Config) (kubernetes.Interface, error) {
 	return kclient, nil
 }
 
-//CRDInstalled to check if the CRD is installed or not
-func CRDInstalled(discovery client.IDiscovery, log logr.Logger) bool {
-	logger := log.WithName("CRDInstalled")
-	check := func(kind string) bool {
-		gvr, err := discovery.GetGVRFromKind(kind)
-		if err != nil {
-			if isServerUnavailable(err) {
-				logger.Info("**WARNING** unable to check CRD status", "kind", kind, "error", err.Error())
-				return true
-			}
-
-			logger.Error(err, "failed to check CRD status", "kind", kind)
-			return false
-		}
-
-		if reflect.DeepEqual(gvr, schema.GroupVersionResource{}) {
-			logger.Info("CRD not installed", "kind", kind)
-			return false
-		}
-		logger.Info("CRD found", "kind", kind)
-		return true
-	}
-
+// CRDsInstalled checks if the Kyverno CRDs are installed or not
+func CRDsInstalled(discovery client.IDiscovery) bool {
 	kyvernoCRDs := []string{"ClusterPolicy", "ClusterPolicyReport", "PolicyReport", "ClusterReportChangeRequest", "ReportChangeRequest"}
 	for _, crd := range kyvernoCRDs {
-		if !check(crd) {
+		if !isCRDInstalled(discovery, crd) {
 			return false
 		}
 	}
+
+	return true
+}
+
+func isCRDInstalled(discoveryClient client.IDiscovery, kind string) bool {
+	gvr, err := discoveryClient.GetGVRFromKind(kind)
+	if gvr.Empty() {
+		if err == nil {
+			err = fmt.Errorf("not found")
+		}
+
+		log.Log.Error(err, "failed to retrieve CRD", "kind", kind)
+		return false
+	}
+
+	log.Log.Info("CRD found", "gvr", gvr.String())
 	return true
 }
 
@@ -192,10 +186,4 @@ func SliceContains(slice []string, values ...string) bool {
 	}
 
 	return false
-}
-
-func isServerUnavailable(err error) bool {
-	// error message -
-	// https://github.com/kubernetes/apimachinery/blob/2456ebdaba229616fab2161a615148884b46644b/pkg/api/errors/errors.go#L432
-	return strings.Contains(err.Error(), "the server is currently unable to handle the request")
 }

